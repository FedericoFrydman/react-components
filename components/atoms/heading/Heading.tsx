import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import './heading.scss';

export type HeadingAs = 'T95' | 'T76' | 'T61' | 'T49' | 'T39' | 'T31' | 'T28' | 'T26' | 'T20' | 'T16' | 'T14';

export interface HeadingProps extends Omit<React.HTMLProps<HTMLHeadingElement>, 'css'> {
  /**
   * Render the text using any HTML element. Default <h2></h2>
   */
  as?: HeadingAs;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The text to be rendered
   */
  text: string;
  /**
   * Use the brand font
   * @deprecated Font is now brand font by default and is determined by `brand.scss`
   */
  useBrand?: boolean;
  /**
   * Defines the title height
   */
  titleHeight?: string;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLHeadingElement>;
}

const HeadingComponent = (
  { as = 'T26', text, titleHeight, classNames = [], useBrand = false, ...props }: HeadingProps,
  ref: HeadingProps['ref'],
) => {
  // Warn for deprecated prop use
  if (useBrand) {
    console.warn('`useBrand` has been deprecated and no longer works.');
  }

  const setHeight = { height: titleHeight, overflow: 'auto' };
  const useHeight = titleHeight ? setHeight : {};

  const classHeading = 'lsux-heading';
  const size = as.substring(1, 3);
  const cnHeading = cn(
    {
      [`${classHeading}`]: true,
      [`${classHeading}--t${size}`]: size,
    },
    classNames,
  );

  switch (as) {
    case 'T95':
      return (
        <h1 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h1>
      );
    case 'T76':
      return (
        <h2 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h2>
      );
    case 'T61':
      return (
        <h3 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h3>
      );
    case 'T49':
      return (
        <h1 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h1>
      );
    case 'T39':
      return (
        <h2 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h2>
      );
    case 'T31':
      return (
        <h3 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h3>
      );
    case 'T28':
      return (
        <h1 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h1>
      );
    case 'T20':
      return (
        <h3 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h3>
      );
    case 'T16':
      return (
        <h4 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h4>
      );
    case 'T14':
      return (
        <h5 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h5>
      );
    case 'T26':
    default:
      return (
        <h2 {...props} className={cnHeading} style={{ ...useHeight, ...props.style }} ref={ref}>
          {text}
        </h2>
      );
  }
};

export const Heading = forwardRef<HTMLHeadingElement, HeadingProps>(HeadingComponent);
