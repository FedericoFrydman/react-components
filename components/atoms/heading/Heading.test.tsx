import { shallow, ShallowWrapper } from 'enzyme';
import { render, screen } from '@testing-library/react';
import { Heading, HeadingAs, HeadingProps } from './Heading';

describe('Heading Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Heading text="some text" />);
    testId = 'test-heading';
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('headingSize', () => {
    const headingSizes: HeadingAs[] = ['T95', 'T76', 'T61', 'T49', 'T39', 'T31', 'T28', 'T26', 'T20', 'T16', 'T14'];
    const text = 'Test my heading size';

    describe('when "headingSize" is defined', () => {
      function itAssignsTheRightClassForHeadingSize(as: HeadingAs) {
        const asClass = `lsux-heading--${as.toLowerCase()}`;
        describe(`when "as" is "${as}"`, () => {
          it(`assigns the element the ${asClass} class`, () => {
            renderHeading({ as, text });

            const textEl = screen.getByText(text);
            expect(textEl).toHaveClass(asClass);
          });
        });
      }

      headingSizes.forEach((headingSize) => itAssignsTheRightClassForHeadingSize(headingSize));
    });

    describe('when titleHeight is defined', () => {
      const text = 'testing text';
      it('is present on the heading element', () => {
        renderHeading({ text, titleHeight: '10px' });
        expect(screen.getByText(text)).toContainHTML('height: 10px;');
      });
    });
  });

  const renderHeading = (props: HeadingProps = { text: 'test' }): HTMLElement => {
    render(<Heading {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
