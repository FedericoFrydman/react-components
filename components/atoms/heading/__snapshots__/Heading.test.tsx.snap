// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading Component should render the component 1`] = `
<h2
  className="lsux-heading lsux-heading--t26"
  style={Object {}}
>
  some text
</h2>
`;
